hdbscan/hdbscan_.py

```diff
@@ -324,20 +398,119 @@ def _hdbscan_boruvka_balltree(X, min_samples=5, alpha=1.0,
-            approx_min_span_tree=True, gen_min_span_tree=False,
-            core_dist_n_jobs=4,
-            cluster_selection_method='eom', allow_single_cluster=False,
-            match_reference_implementation=False, **kwargs):
+def remap_condensed_tree(tree, internal_to_raw, outliers):
+    """
+    Takes an internal condensed_tree structure and adds back in a set of points
+    that were initially detected as non-finite and returns that new tree.
+    These points will all be split off from the maximal node at lambda zero and
+    considered noise points.
+
+    Parameters
+    ----------
+    tree: condensed_tree
+    internal_to_raw: dict
+        a mapping from internal integer index to the raw integer index
+    finite_index: ndarray
+        Boolean array of which entries in the raw data were finite
+    """
+    finite_count = len(internal_to_raw)
+
+    outlier_count = len(outliers)
+    for i, (parent, child, lambda_val, child_size) in enumerate(tree):
+        if child < finite_count:
+            child = internal_to_raw[child]
+        else:
+            child = child + outlier_count
+        tree[i] = (parent + outlier_count, child, lambda_val, child_size)
+
+    root = tree[0][0]  # Should I check to be sure this is the minimal lambda?
+    for i in outliers:
+        new_leaf = np.array(
+            (root, i, 0, 1),
+            dtype=[
+                ("parent", np.intp),
+                ("child", np.intp),
+                ("lambda_val", float),
+                ("child_size", np.intp),
+            ],
+        )
+        tree = np.append(new_leaf, tree)
+    return tree
+
+
+def remap_single_linkage_tree(tree, internal_to_raw, outliers):
+    """
+    Takes an internal single_linkage_tree structure and adds back in a set of points
+    that were initially detected as non-finite and returns that new tree.
+    These points will all be merged into the final node at np.inf distance and
+    considered noise points.
+
+    Parameters
+    ----------
+    tree: single_linkage_tree
+    internal_to_raw: dict
+        a mapping from internal integer index to the raw integer index
+    finite_index: ndarray
+        Boolean array of which entries in the raw data were finite
+    """
+    finite_count = len(internal_to_raw)
+
+    outlier_count = len(outliers)
+    for i, (left, right, distance, size) in enumerate(tree):
+        if left < finite_count:
+            tree[i, 0] = internal_to_raw[left]
+        else:
+            tree[i, 0] = left + outlier_count
+        if right < finite_count:
+            tree[i, 1] = internal_to_raw[right]
+        else:
+            tree[i, 1] = right + outlier_count
+
+    for i in outliers:
+        last_cluster_id = tree[tree.shape[0] - 1][0:2].max()
+        last_cluster_size = tree[tree.shape[0] - 1][3]
+        new_leaf = np.array((i, last_cluster_id + 1, np.inf, last_cluster_size + 1))
+        tree = np.append(tree, [new_leaf], axis=0)
```

Comment:
Same as with the condensed tree: with a lot of outliers this could theoretically get quite expensive with all these appends (which do a memory reallocation each time).